### ISSUE TITLE
Groundwork for Multidimensional Gas Fees

### ISSUE THREAD

**Issue by hdevalence** [penumbra-zone/penumbra#2970]
**Is your feature request related to a problem? Please describe.**

Processing transactions consumes resources. To account for this, transactions should pay fees. Currently, we've designed a way for Penumbra transactions to include a fee payment, but we haven't designed the system that controls what those fees should be.

Penumbra transactions can consume different types of resources. For instance, proof verifications are parallelizable and purely CPU-bound, while IBC packets may require sequential, stateful accesses. Most importantly, new output notes create work not just for every _node_, but every _user_, who must scan the new notes to determine whether they are relevant (or offload that scanning work to a server, using fuzzy message detection).

These resources have different capacity limits, and those capacity limits show themselves at different time scales. For instance, Zcash experienced a DoS attack where an adversary posted many transactions that included many outputs. While these transactions didn't exceed the available block space, they did overwhelm other clients' effective scanning capacity, cause all other users to have degraded UX.  (See [this discussion](https://twitter.com/hdevalence/status/1597115071033118722) for details).

In Penumbra's context, we can express the problem as follows: while _block space_ describes synchronization bandwidth between full nodes, _compact block space_ describes synchronization bandwith among client devices. On long time scales, compact block space is much more constrained than block space, because client devices are assumed to have much more restrictive bandwidth constraints than full nodes, but on short time scales, there's no particular constraint -- a client trying to synchronize the last week's worth of activity doesn't particularly care about whether it was evenly spread or concentrated in a few bursts.

So the resources we need to account for differ not just in their capacity but also in their time horizons. How can we handle this? One solution is provided by the [multidimensional resource pricing](https://baincapitalcrypto.com/multi-dimensional-on-chain-resource-pricing/) paper, which shows why multidimensional gas fees are better than a scalar gas quantity, and how to automatically price different resources according to actual utilization.

**Describe the solution you'd like**

We should implement this for Penumbra, in a phased approach that lets us frontload foundational structure and backload whatever implementation complexity we can get away with.  I think we can decompose the work into three phases:

_Resource Selection_
 
First, what resources do we need to price? Based on discussions with the team, it seems like the following is a reasonable starting point:

1. **Block Space**, a proxy for synchronization bandwidth between full nodes.
2. **Compact Block Space**, a proxy for synchronization bandwidth for clients.
3. **Verification**, a proxy for parallelizable, CPU-bound work.
4. **Execution**, a proxy for non-parallelizable stateful execution.

_Gas Schedules_

In the `transaction` crate:

- [x] Define a `Gas` structure to represent a multidimensional gas vector.

This should look like
```rust
struct Gas {
    pub block_space: u64,
    pub compact_block_space: u64,
    pub verification: u64,
    pub execution: u64,
}
```
It should have an `Add` impl to allow adding gas costs.

- [x] Define a `GasCost` trait that assigns a gas cost to each `Action`.

Making this a standalone trait means we can put all the impls in one place, so we have one file declaring the entire gas cost schedule.  We should make the gas costs statically determined, because that's easier for clients to reason about, and we should avoid allowing transactions to do arbitrary work on-chain anyways, so static gas costs are an additional forcing function.

The `GasCost` trait should be implemented for each `Action` and for the `Transaction` overall.  The `Transaction` implementation should sum up all of the constituent actions' gas costs and add its own contribution. This way, we can have each `Action`'s implementation have `block_space: 0` and handle the block space in the `Transaction` impl by using the encoded size of the transaction. (Yes, this means wastefully re-encoding. We can deal with that if and when it shows up in a profile).

For the gas schedule itself, we can just make up some numbers and fine tune them later. Because gas needs to be multiplied by prices, only the relative costs of different actions matter, the rest gets factored into the prices. As a rough starting point, what about:

- Use the encoded size (in bytes) of whatever data the action adds to the compact block as the compact block space;
- Use `1000` as the `verification` cost for anything with a snark and `0` otherwise;
- Use `10` for `execution` cost.

Later, we can fine-tune these.

- [x] Define a `GasPrices` proto and domain type 

This should also be a four-dimensional vector, but instead it expresses the price of each unit of gas in terms of the staking token:
```rust
struct GasPrices {
    pub block_space_price: u64,
    pub compact_block_space_price: u64,
    pub verification_price: u64,
    pub execution_price: u64,
}

impl GasPrices {
    pub fn price(&self, gas: &Gas) -> Amount {
        // dot product
    }
}
```

- [x] Add hardcoded gas prices as a chain parameter

To start, we can hardcode the gas prices as a chain parameter, and have the default parameter choice be all-zero. This lets us experiment with nonzero fees on devnets (by editing the generated parameters in the `genesis.json`) prior to committing to the change and potentially breaking client software.

- [x] Add gas checks to the transaction-level fee logic.

The chain logic should enforce that the transaction-level fee is greater than or equal to the _base fee_, which is computed as the transaction's gas use, priced by the current gas prices. (Initially, gas prices are static, so the static prices are always current).  For now, we can continue to do nothing with the fee, effectively burning it, though we may want to revisit this later once there's a working distributions component (at which point we can distribute everything above the base fee to the proposer, giving an in-protocol first-price auction for block space).

- [x] Add an RPC for fetching the current gas prices.

This just amounts to adding the GRPC method to the `SpecificQuery` RPC; the initial implementation can return the hardcoded chain parameters. Although the implementation is trivial, adding it as an RPC sets the expectation that client software should query for gas prices (so that when we make them dynamic later, no changes are required).

- [x] Change the `TransactionPlanner` to account for fees.

This means the `TransactionPlanner` will need to be configured with a `GasPrices` at the beginning. If none is supplied, it should internally set an all-zero gas prices (to retain compatibility with current API use).

We want to retain the existing two-phase planner, which is convenient for use in the wasm context where the planner code may not have state access.  In the current planner API, the caller specifies a list of constraints on a planner object, and the planner object returns information about what it needs to balance the transaction (i.e., what notes it needs to be able to spend).

This potentially poses a problem: if the planner estimates gas based on the constraints, but those constraints are satisfied by adding additional actions, the resulting transaction may underpay fees. One can imagine pathological cases where this cascades over multiple rounds, etc.

As a workaround that would be much simpler and probably work in 99% of cases is to just "overbudget" the estimated gas based on the provided constraints, and then adjust it into change outputs according to the actual gas use.  For instance, we could multiply the estimated gas by 2x or 3x, request funds to balance the transaction with overpaid gas, and then refund any excess into the transaction's existing change output in the second phase of planning.  (In particular, we expect that outputs are much more expensive than spends, since they consume >5x the compact block space, so satisfying funding constraints should be much less expensive than adding them).

**Followup work**

At this point, we've laid the groundwork for multidimensional gas prices, but ensured that the prices are all set to zero (except potentially in a local devnet with customized parameters). This means that we don't need to change anything about the client code for it to keep working.

In a later phase, we should follow on this groundwork:

- [ ] #2971
- [ ] #2972

**Comment by conorsch** [comment]
We'll tackle this one for Testnet 61. The follow-up work identified at the bottom of the ticket likely won't make 61, and that's OK.

**Comment by hdevalence** [comment]
This is blocked on #3084 , in order to put the prices in the chain parameters.
